File: surfaces/ruby/lib/cleanpaste_normalizer.rb
module CleanPasteNormalizer
  ZERO_WIDTH = "\u200B\u200C\u200D\u200E\u200F\u202A\u202B\u202C\u202D\u202E\u2060\uFEFF"
  UNICODE_SPACES = "\u00A0\u1680\u2000\u2001\u2002\u2003\u2004\u2005\u2006\u2007\u2008\u2009\u200A\u202F\u205F\u3000"

  def self.normalize_pasted_text(text)
    raise TypeError, "text must be a string" unless text.is_a?(String)

    normalized = text.gsub("\r\n", "\n").gsub("\r", "\n")
    normalized = normalized.unicode_normalize(:nfkc)
    normalized = normalized.delete(ZERO_WIDTH).tr(UNICODE_SPACES, " " * UNICODE_SPACES.length)
    normalized = normalized.gsub(/[ \t]+\n/, "\n")
    normalized.sub(/[ \t]+\z/, "")
  end
end

File: surfaces/dart/lib/cleanpaste_normalizer.dart
final RegExp _zeroWidth = RegExp(
  r'[\u200b\u200c\u200d\u200e\u200f\u202a\u202b\u202c\u202d\u202e\u2060\ufeff]',
);
final RegExp _unicodeSpaces = RegExp(
  r'[\u00a0\u1680\u2000-\u200a\u202f\u205f\u3000]',
);

String normalizePastedText(String text) {
  var normalized = text.replaceAll('\r\n', '\n').replaceAll('\r', '\n');
  normalized = _nfkcCompatibility(normalized)
      .replaceAll(_zeroWidth, '')
      .replaceAllMapped(_unicodeSpaces, (_) => ' ')
      .replaceAll(RegExp(r'[ \t]+\n'), '\n')
      .replaceFirst(RegExp(r'[ \t]+$'), '');
  return normalized;
}

String _nfkcCompatibility(String text) => String.fromCharCodes(
      text.runes.map((rune) {
        if (rune >= 0xFF01 && rune <= 0xFF5E) return rune - 0xFEE0;
        if (rune == 0x3000) return 0x20;
        return rune;
      }),
    );
